### PR TITLE
make bot turn calculations more tolerant of threading-related errors fixable by kicking and replacing the bot

### LINKED ISSUE #2141: [0.47.9] ConcurrentModificationException while calculating bot firing turn
### Environment
MekHQ integrated Megamek - v0.47.9
Windows 10
Java version 1.8.0_261

### Description
Princess bot stalled due to a ConcurrentModificationException while calculating fire for a unit. See attached logs and savegame for details. 

Edit: After further checking, loading the savegame does not reproduce the bug on my end, so sadly it probably won't on yours either. Hopefully the logs are helpful. 

### Files
[megameklog.txt](https://github.com/MegaMek/megamek/files/5066284/megameklog.txt)
[mekhqlog.txt](https://github.com/MegaMek/megamek/files/5066286/mekhqlog.txt)
[indirectFireBug.sav.gz](https://github.com/MegaMek/megamek/files/5066304/indirectFireBug.sav.gz)
Player name: Coppercoats

### PR DESCRIPTION
**Pull request by NickAragua**
Theoretically fixes #2141 

I'm guessing this is caused when the bot is calculating a firing turn and receives an update for the entity being examined. Why or how this would happen, I don't know.

Not really sure what the impact will be on performance, or if this the right way of going about it (will this change actually prevent the concurrent modification exception?)

### CODE REVIEW

File: megamek/src/megamek/common/EquipmentType.java [hunk 1]
```diff
@@ -553,10 +555,16 @@ public boolean hasModeType(String modeType) {
     		return false;
     	}
     	
-    	for(EquipmentMode mode : modes) {
-    		if(mode.getName().equals(modeType)) {
-    			return true;
-    		}
+    	// sometimes, the list of modes may be modified while we're looping through it
+    	// so let's do this in a thread-safe way
+    	List<EquipmentMode> threadSafeModes = Collections.synchronizedList(modes);
```
Comment (sixlettervariables) [inline megamek/src/megamek/common/EquipmentType.java:560]: `Collections.synchronizedList` protects the given methods on `List` from concurrent access, however, they do not protect iteration. So you'd need to use `synchronized`.

Except, `threadSafeModes` isn't being used for all of your mutable uses of `modes` (`add`, `remove`). As it stands, you'd likely be better served by simply wrapping every usage of `modes` in a `synchronized` block on `modes` itself.

Comment (sixlettervariables) [reply megamek/src/megamek/common/EquipmentType.java:560]: Also it looks like `modes` is a `Vector`, which already has the same behavior as a `synchronizedList`. So basically, either `addMode` or `removeMode` was called while this was going on. My guess is a network delay or GC pause gave it enough timing to fail.

**Comment by NickAragua** [comment]
Went with a different approach - now, the two areas of bot code that exhibit crashes (as described in #2141 and #2142) which can be resolved by kicking and replacing the bot have a retry mechanism. The methods will attempt to retry up to three times, waiting between 500 and 1500 milliseconds between retries.

**Comment by NickAragua** [comment]
Also fixed an observed bug during testing where a spheroid dropship in atmosphere that was out of control appeared to allow the player to move it.

File: megamek/src/megamek/client/bot/princess/PathEnumerator.java [hunk 2]
```diff
@@ -168,166 +170,173 @@ void updateUnitLocations(Entity entity, List<MovePath> paths) {
...
-            getOwner().methodEnd(getClass(), METHOD_NAME);
+        
+        int retryCount = 0;
+        boolean success = false;
+        
+        while ((retryCount < BotClient.BOT_TURN_RETRY_COUNT) && !success) {
+	        try {
+	
+	            // Record it's current position.
+	            getLastKnownLocations().put(
```
Comment (sixlettervariables) [inline megamek/src/megamek/client/bot/princess/PathEnumerator.java:181]: Same as above.

**Comment by NickAragua** [comment]
Refactored. Found another bug while I was in there (seems that superheavy mechs can't unjam rotary autocannons for some reason), but I'll deal with that one separately.

**Pull request merged**